Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,591 @@
+                return false
+            }
+        }
+
+        /// The connection is active, but there are no running requests on the connection.
+        /// Every idle connection is available, but not every available connection is idle.
+        var isIdle: Bool {
+            switch self.state {
+            case .active(_, _, let usedStreams, _):
+                return usedStreams == 0
+            case .starting, .backingOff, .draining, .closed:
+                return false
+            }
+        }
+
+        var isClosed: Bool {
+            switch self.state {
+            case .starting, .backingOff, .draining, .active:
+                return false
+            case .closed:
+                return true
+            }
+        }
+
+        private var state: State
+        let eventLoop: EventLoop
+        let connectionID: Connection.ID
+
+        /// should be called after the connection was successfully established
+        /// - Parameters:
+        ///   - conn: HTTP2 connection
+        ///   - maxStreams: max streams settings from the server
+        /// - Returns: number of available streams which can be leased
+        mutating func connected(_ conn: Connection, maxStreams: Int) -> Int {
+            switch self.state {
+            case .active, .draining, .backingOff, .closed:
+                preconditionFailure("Invalid state: \(self.state)")
+
+            case .starting:
+                self.state = .active(conn, maxStreams: maxStreams, usedStreams: 0, lastIdle: .now())
+                return maxStreams
+            }
+        }
+
+        /// should be called after receiving new http2 settings from the server
+        /// - Parameters:
+        ///   - maxStreams: max streams settings from the server
+        /// - Returns: number of available streams which can be leased
+        mutating func newMaxConcurrentStreams(_ maxStreams: Int) -> Int {
+            switch self.state {
+            case .starting, .backingOff, .closed:
+                preconditionFailure("Invalid state for updating max concurrent streams: \(self.state)")
+
+            case .active(let conn, _, let usedStreams, let lastIdle):
+                self.state = .active(conn, maxStreams: maxStreams, usedStreams: usedStreams, lastIdle: lastIdle)
+                return max(maxStreams - usedStreams, 0)
+
+            case .draining(let conn, _, let usedStreams):
+                self.state = .draining(conn, maxStreams: maxStreams, usedStreams: usedStreams)
+                return 0
+            }
+        }
+
+        mutating func goAwayReceived() -> EventLoop {
+            switch self.state {
+            case .starting, .backingOff, .closed:
+                preconditionFailure("Invalid state for draining a connection: \(self.state)")
+
+            case .active(let conn, let maxStreams, let usedStreams, _):
+                self.state = .draining(conn, maxStreams: maxStreams, usedStreams: usedStreams)
+                return conn.eventLoop
```

Comment:
AFAIK, the currently planned strategy was to not close any streams but just not accept any new request on that connection. @fabianfett please correct me if I'm wrong.

Reply:
That makes sense, but GOAWAY itself closes streams. The GOAWAY frame contains a Last-Stream-ID field: streams with stream IDs above this number are treated as if they had never existed, and so are immediately and irrevocably closed. SwiftNIO HTTP/2 will inform you of this.

Reply:
@Lukasa the `HTTP2IdleHandler` takes care of all this logic and is the source for the GOAWAY received call.
https://github.com/swift-server/async-http-client/blob/main/Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2IdleHandler.swift 

The idea is, we can't close the connection right away, after the GOAWAY. So we try to inform the ConnectionPool as early as possible. When we have no open streams in `HTTP2IdleHandler`, we will close the `HTTP2Connection` and inform the delegate (the connection pool) about it.

Reply:
Because of the ping pong between handler and multiplexer, we observe here the intent of GOAWAY. And we will notice the stream close based on `StreamCloseEvents`. 

https://github.com/swift-server/async-http-client/blob/316cbf935353c3c138d7627c5f5d9a7784181270/Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2IdleHandler.swift#L84

In the pool we only consume the infos that are created in HTTP2IdleHandler.

Reply:
Great, that's what I figured, but I did want to make sure.